### PR TITLE
[WIP] On-screen keyboard support

### PR DESCRIPTION
**Pull request by dcz-purism**
This pull request combines work on:
- `input-method` protocol #856 
- `text-input` protocol #776 
- rootston glue

Its purpose is to develop these in sync, and to remove the need to juggle different long-running feature branches as different parts are being developed. This has proven to take up effort and lead to awkward situations, e.g. many of the rebased rootston commits were developed against an earlier version of `wlr_input_method.h` and are currently unbuildable.

One interesting thing is the addition of code duplicating some functionality of layer surfaces (panels, unfocusable windows) in rootston views, in order to support `input-method-panel`.

### CODE REVIEW

File: include/rootston/osk.c [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	if (wl_resource == NULL) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+	wl_resource_set_implementation(wl_resource, &input_panel_impl, NULL, NULL);
+}
+
+struct roots_input_panel *roots_input_panel_create(struct wl_display *display) {
+	struct roots_input_panel *panel = calloc(1,
+											 sizeof(struct roots_input_panel));
```
Comment (SirCmpwn) [inline include/rootston/osk.c]: Bad indent

File: include/rootston/osk.c [hunk 2]
```diff
@@ -0,0 +1,53 @@
+#include "osk.h"
+
+#include "input-method-unstable-v1-protocol.h"
+
+static void noop() {}
```
Comment (SirCmpwn) [inline include/rootston/osk.c]: Add newlines and such and something to the effect of `/* This space deliberately left blank */`

File: include/rootston/osk.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+#include <stdlib.h>
+#include <stdint.h>
+#include <assert.h>
+
+#include <wayland-server-core.h>
+
+#include "rootston/desktop.h"
+
+struct roots_input_panel {
+	struct roots_desktop *desktop; // FIXME - remove when views come
```
Comment (SirCmpwn) [inline include/rootston/osk.h]: If you expect this TODO to get merged, can you elaborate a bit? If you will resolve prior to merge no worries

**Comment by agx** [comment]
Now that we have layer-surfaces can't OSK use just that?

File: include/rootston/text_input.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+#include <wlr/types/wlr_input_method.h>
+#include <wlr/types/wlr_text_input.h>
+#include "rootston/seat.h"
+
+struct roots_text_input {
+	struct roots_seat *seat;
+    struct wlr_text_input *input;
+	struct wlr_input_method_context *context;
+
+    struct wl_list link;
```
Comment (SirCmpwn) [inline include/rootston/text_input.h]: Wrong indent here and sprinkled throughout this struct

File: include/rootston/text_input.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+	struct wl_listener text_input_enable; // uint32_t show_input_panel
```
Comment (SirCmpwn) [inline include/rootston/text_input.h]: Put the docs near the signal, not the listener

File: include/rootston/view.h [hunk 6]
```diff
@@ -70,13 +71,27 @@ struct roots_xwayland_surface {
...
+};
+
 enum roots_view_type {
 	ROOTS_WL_SHELL_VIEW,
 	ROOTS_XDG_SHELL_V6_VIEW,
 	ROOTS_XDG_SHELL_VIEW,
 #ifdef WLR_HAS_XWAYLAND
 	ROOTS_XWAYLAND_VIEW,
 #endif
+	ROOTS_INPUT_PANEL_VIEW,
```
Comment (SirCmpwn) [inline include/rootston/view.h]: This should probably not be a view, but an entirely different abstraction (like layer shell does)

File: include/rootston/view.h [hunk 7]
```diff
@@ -92,6 +107,13 @@ struct roots_view {
 	int border_width;
 	int titlebar_height;
 
+	bool special; // marks special "system" feature set
```
Comment (SirCmpwn) [inline include/rootston/view.h]: gross

File: include/wlr/types/wlr_input_method.h [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+		struct wl_signal destroy; // struct wlr_input_panel_surface*
+	} events;
+};
+
+struct wlr_input_method {
+	struct wl_global *input_method;
+	struct wl_resource *resource;
+	struct {
+		struct wl_signal new_context; // struct wlr_input_method_context*
+		// TODO: create roots_keyboard (seat_add_keyboard) on this; modify roots_keyboard to support backend-less
```
Comment (SirCmpwn) [inline include/wlr/types/wlr_input_method.h]: I think this should be more along the lines of wlr_input_method creating a wlr_keyboard which roots_keyboard should just be able to pick up without changes. Maybe some silly changes to automatically associate it with the relevant seat if `wlr_keyboard_is_input_method` returns true.

File: include/wlr/types/wlr_input_method.h [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+struct wlr_input_method_context *wlr_input_method_send_activate(
+	struct wlr_input_method *input_method);
+
+void wlr_input_method_context_send_surrounding_text(
+	struct wlr_input_method_context *input_method, const char *text,
+	uint32_t cursor, uint32_t anchor);
+
+void wlr_input_method_send_deactivate(struct wlr_input_method *input_method,
+	struct wlr_input_method_context *context);
+#endif
```
Comment (SirCmpwn) [inline include/wlr/types/wlr_input_method.h]: minor nit: extra newline above the `#endif`

**Comment by dcz-purism** [comment]
@agx: I wanted to have a reasonably complete implementation of the existing protocol first, especially because all existing software, namely weston-keyboard uses it. The plan is to update the protocol and remove the panel in its current form in favor of layer-surfaces in the future.

File: rootston/cursor.c [hunk 10]
```diff
@@ -273,7 +275,9 @@ static void roots_cursor_press_button(struct roots_cursor *cursor,
 			}
 			break;
 		case WLR_BUTTON_PRESSED:
-			roots_seat_set_focus(seat, view);
+			if (!view || !view->special || view->features.focusable) {
+				roots_seat_set_focus(seat, view);
+            }
```
Comment (SirCmpwn) [inline rootston/cursor.c]: Weird indent

File: rootston/desktop.c [hunk 11]
```diff
@@ -535,6 +552,33 @@ void view_update_position(struct roots_view *view, double x, double y) {
...
+		break;
+	default:
+		break;
+	}
+
+	view_damage_whole(view);
+	view->x = x;
+	view->y = y;
+	view_damage_whole(view);
+}
```
Comment (SirCmpwn) [inline rootston/desktop.c]: Yeah all of this stuff can go if you don't try to hack it into views.

File: rootston/input.c [hunk 12]
```diff
@@ -43,7 +43,7 @@ struct roots_seat *input_get_seat(struct roots_input *input, char *name) {
 	return seat;
 }
 
-static void handle_new_input(struct wl_listener *listener, void *data) {
+void handle_new_input(struct wl_listener *listener, void *data) {
```
Comment (SirCmpwn) [inline rootston/input.c]: uh

**Comment by agx** [comment]
@dcz-purism with existing you mean weston-keyboard using zwp_input_panel_surface_v1 ?

File: rootston/output.c [hunk 13]
```diff
@@ -333,6 +336,9 @@ static bool has_standalone_surface(struct roots_view *view) {
 	case ROOTS_XWAYLAND_VIEW:
 		return wl_list_empty(&view->xwayland_surface->children);
 #endif
+	case ROOTS_INPUT_PANEL_VIEW:
+		printf("Stub: has_standalone_surface on OSK_VIEV returns true\n");
+		return true;
```
Comment (Shugyousha) [inline rootston/output.c]: The printed text should be updated to correspond to the constant. There is also a typo in the current text (s/OSK_VIEV/OSK_VIEW/).

File: rootston/seat.c [hunk 14]
```diff
@@ -905,9 +932,13 @@ void roots_seat_cycle_focus(struct roots_seat *seat) {
 		return;
 	}
 
-	// Focus the next view
-	struct roots_seat_view *next_seat_view = wl_container_of(
-		first_seat_view->link.next, next_seat_view, link);
+	// Focus the next view which is focusable
+	struct roots_seat_view *next_seat_view = first_seat_view;;
+	do {
```
Comment (Shugyousha) [inline rootston/seat.c]: double ;

File: include/rootston/view.h (hunk 7 above)
Comment (dcz-purism) [reply include/rootston/view.h]: ugly - this code will go when panel interface is reconstructed, so I went crazy here.

File: types/wlr_text_input.c [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+	// no need to destroy the resource if client closed, prompting the server to do it
+	if (text_input->resource) {
+		wl_resource_destroy(text_input->resource);
+	}
+	if (text_input->current.surrounding.text) {
+		free(text_input->current.surrounding.text);
+	}
+	if (text_input->pending.surrounding.text) {
+		free(text_input->pending.surrounding.text);
+	}
```
Comment (Shugyousha) [inline types/wlr_text_input.c]: Since free(NULL) is a noop, we don't need these two "if" statements here.

File: types/wlr_text_input.c [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+}
+
+static void text_input_set_surrounding_text(struct wl_client *client,
+		struct wl_resource *resource, const char *text, int32_t cursor,
+		int32_t anchor) {
+	struct wlr_text_input *text_input = text_input_from_resource(resource);
+	if (text_input->pending.surrounding.text) {
+		free(text_input->pending.surrounding.text);
+	}
+	text_input->pending.surrounding.text = strdup(text);
```
Comment (Shugyousha) [inline types/wlr_text_input.c]: Same here

File: types/wlr_text_input.c [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+	text_input->pending.cursor_rectangle.height = height;
+}
+
+static void text_input_commit(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_text_input *text_input = text_input_from_resource(resource);
+	if (text_input->current.surrounding.text) {
+		free(text_input->pending.surrounding.text);
+	}
+	text_input->current = text_input->pending;
```
Comment (Shugyousha) [inline types/wlr_text_input.c]: Ditto

File: types/wlr_text_input.c [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+		return;
+	}
+	wl_resource_set_implementation(wl_resource, &text_input_manager_impl,
+		manager, NULL);
+}
+
+struct wlr_text_input_manager *wlr_text_input_manager_create(
+		struct wl_display *wl_display) {
+	struct wlr_text_input_manager *manager =
+		calloc(1, sizeof(struct wlr_text_input_manager));
```
Comment (Shugyousha) [inline types/wlr_text_input.c]: Add out of memory check

**Comment by agx** [comment]
@dcz-purism my main concern is growing a third way to handle popups and overlay surfaces. layer-shell already has overlap with views and it would be nice if either one could be reused.

File: rootston/output.c (hunk 13 above)
Comment (dcz-purism) [reply rootston/output.c]: What's a standalone surface here?

Comment (emersion) [reply rootston/output.c]: It's used for fullscreen to know if the view has only one surface.

File: examples/text-input.c [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+#include <wayland-egl.h>
+#include <wlr/render/egl.h>
+#include "text-input-unstable-v3-client-protocol.h"
+#include "xdg-shell-client-protocol.h"
+
+#include <linux/input-event-codes.h>
+
+/**
+ * Usage: idle-inhibit
+ * Creates a xdg-toplevel using the idle-inhibit protocol.
```
Comment (agx) [inline examples/text-input.c]: It's rather text-input and does s.th. different?

File: rootston/input_method.c [hunk 20]
```diff
@@ -134,3 +144,55 @@ void handle_input_method_context_destroy(struct wl_listener *listener,
 	}
 	update_maximized_views(desktop);
 }
+
+static void spawn_im_process() {
+	*im_pid = fork();
+	if (*im_pid == 0) {
+		sleep(1);
+		execl("/bin/sh", "/bin/sh", "-c", im_cmd, (void *)NULL);
```
Comment (agx) [inline rootston/input_method.c]: Why does it need a shell?

**Comment by agx** [comment]
On Thu, Apr 26, 2018 at 09:22:57PM +0000, dcz-purism wrote:
> dcz-purism commented on this pull request.
> 
> 
> 
> > @@ -134,3 +144,55 @@ void handle_input_method_context_destroy(struct wl_listener *listener,
>  	}
>  	update_maximized_views(desktop);
>  }
> +
> +static void spawn_im_process() {
> +	*im_pid = fork();
> +	if (*im_pid == 0) {
> +		sleep(1);
> +		execl("/bin/sh", "/bin/sh", "-c", im_cmd, (void *)NULL);
> 
> The -E command accepts a shell, so in order to maintain consistency, I
> made it the same. Personally, I don't see the need for a shell in
> either case, but maybe I'm missing something.

Shell uses $PATH but I'd rather not rely on that for the OSK.

**Comment by SirCmpwn** [comment]
How're are things in this department going on your end? Should we expect an update to this soon?

**Comment by dcz-purism** [comment]
Yes indeed! Currently I'm busy upstreaming *text-input* to GTK, so after the merge window closes and I'm free to do other things again, I'll push my changes to get some formal feedback.

**Comment by SirCmpwn** [comment]
Sweet, keep up the good work!

**Comment by dcz-purism** [comment]
The first replacement of the old input method was just posted to gather feedback today:

https://lists.freedesktop.org/archives/wayland-devel/2018-August/039255.html

**Comment by dcz-purism** [comment]
I rewrote this code to updated protocols and submitted it as a separate (out of forgetfulness) PR [here](https://github.com/swaywm/wlroots/pull/1203).

**Pull request closed without merging**